Initial loongarch port

xiangzhai opened the pull request:
Hi,

I want to port vectorscan AKA [hyperscan](https://github.com/intel/hyperscan) for [LoongArch Architecture](https://docs.kernel.org/loongarch/index.html).

Passed unit-internal test:

```
[==========] 14040 tests from 73 test cases ran.
[  PASSED  ] 14040 tests.
```

Passed unit-hyperscan test:

```
[==========] 3746 tests from 33 test cases ran.
[  PASSED  ] 3746 tests.
```

Thanks,
Leslie Zhai

hongyang7 committed:
literal API: add empty string check.

fixes github issue #302, #304

hongyang7 committed:
literal API: add instruction support

fixes github issue #303

hongyang7 committed:
mcclellan: improve wide-state checking in Sherman optimization

fixes github issue #305

fatchanghao committed:
Logical Combination: bypass combination flag in hs_expression_info.

Fixes github issue #291

hongyang7 committed:
update year for bugfix #302-#305

hongyang7 committed:
bugfix: add vbmi platform parameter for tests in single.cpp

hongyang7 committed:
bugfix: add vbmi case for test in database.cpp

markos committed:
update Jenkinsfile for all configurations

markos committed:
lower gcc minver to 9 to enable building on Ubuntu 20 LTS

markos committed:
fix wrong castings for NEON

markos committed:
fix clang-release-arm compilation

markos committed:
minor fixes

markos committed:
Merge pull request #81 from VectorCamp/feature/add-clang-support

Feature/add clang support

markos committed:
Minor changes to enable compilation on Mac M1

markos committed:
Merge pull request #82 from VectorCamp/feature/add-macos-support

Minor changes to enable compilation on Mac M1

hongyang7 committed:
Fix segfaults on allocation failure (#4)

Throw std::bad_alloc instead of returning nullptr from
ue2::AlignedAllocator. Allocators for STL containers are expected never
to return with an invalid pointer, and instead must throw on failure.
Violating this expectation can lead to invalid pointer dereferences.

Co-authored-by: johanngan <johanngan.us@gmail.com>

fixes github issue #317 (PR #320)

markos committed:
Initial attempt at debian packaging, modified hyperscan packaging

markos committed:
fix DEB_CMAKE_FLAGS depending on DEB_HOST_ARCH

hongyang7 committed:
sanitiser bugfix

BigRedEye committed:
fix: Mark operator bool explicit

markos committed:
enable FAT_RUNTIME

markos committed:
defer setting arch/tune flags for FAT_RUNTIME

markos committed:
add our copyrights, minor fixes

markos committed:
remove preinst script, not needed as we bumped our deps

markos committed:
add sse4.2-support package to enforce such dependency

markos committed:
add ITP bug report

markos committed:
fix description, remove sse4.2-support from b-depends

markos committed:
change source format to native, as we include debian folder

markos committed:
Merge pull request #85 from VectorCamp/feature/add-debian-package

Feature/add debian package

markos committed:
keep debian folder in a separate branch

markos committed:
Merge pull request #87 from VectorCamp/feature/move-debian-package-to-separate-branch

keep debian folder in a separate branch

markos committed:
Merge pull request #86 from VectorCamp/develop

New release 5.4.6

markos committed:
Merge pull request #90 from BigRedEye/vectorscan-master

Fix word boundary assertions under C++20

danilak-G committed:
Fix all ASAN issues in vectorscan

danilak-G committed:
Add sanitize options

danilak-G committed:
Fix a couple of tests

a16bitsysop committed:
change FAT_RUNTIME to a normal option so it can be set to off

fixes #89

a16bitsysop committed:
move to original position

markos committed:
Merge pull request #93 from danlark1/master

Fix all ASAN issues in vectorscan

markos committed:
Merge pull request #94 from a16bitsysop/fat_runtime

change FAT_RUNTIME to a normal option so it can be set to off

danlark1 committed:
Optimized and correct version of movemask128 for ARM

Closes #99

https://gcc.godbolt.org/z/cTjKqzcvn

Previous version was not correct because movemask thought of having bytes 0xFF. We can fully match the semantics + do it faster with USRA instructions.

Re-submission to a develop branch

markos committed:
Merge pull request #102 from danlark1/patch-2

Optimized and correct version of movemask128 for ARM

markos committed:
add Jenkinsfile back to master branch

markos committed:
add Jenkinsfile back to master branch

hongyang7 committed:
bugfix: fix overflow risk of strlen function

markos committed:
Merge pull request #104 from VectorCamp/bugfix/jenkinsfile

add Jenkinsfile back to master branch

markos committed:
Delete JenkinsFile

markos committed:
fix large pipeline error

markos committed:
Merge pull request #105 from VectorCamp/bugfix/jenkins

fix large pipeline error

markos committed:
Update Jenkinsfile

markos committed:
Update Jenkinsfile

markos committed:
Update Jenkinsfile

markos committed:
Merge pull request #103 from VectorCamp/develop

Develop

markos committed:
Update CMakeLists.txt

fatchanghao committed:
Corpus generator: fix random char value of UTF-8.

fixes github issue #184

fatchanghao committed:
Corpus editor: fix random char value of UTF-8.

jth committed:
Use non-deprecated method of finding python

jth committed:
Bump scripts to python3

markos committed:
Merge pull request #108 from jth/cmake-python

CMake: Use non-deprecated method for finding python

hongyang7 committed:
klocwork: fix risk issues

danilak-G committed:
Optimize vectorscan for aarch64 by using shrn instruction

This optimization is based on the thread
https://twitter.com/Danlark1/status/1539344279268691970 and uses
shift right and narrow by 4 instruction https://developer.arm.com/documentation/ddi0596/2020-12/SIMD-FP-Instructions/SHRN--SHRN2--Shift-Right-Narrow--immediate--

To achieve that, I needed to redesign a little movemask into comparemask
and have an additional step towards mask iteration. Our benchmarks
showed 10-15% improvement on average for long matches.

danilak-G committed:
Fix formatting of a couple files

danilak-G committed:
Minor fix

danilak-G committed:
Fix ppc64el debug

hdu-sdlzx committed:
fix build with glibc-2.34

SIGTSKSZ is no long a constant after glibc 2.34
https://sourceware.org/pipermail/libc-alpha/2021-August/129718.html

hongyang7 committed:
gcc-10(and above): fix compile issue caused by stringop-overflow

hongyang7 committed:
stringop-overflow compatible fix

danlark1 committed:
Fix avx512 movemask call

Commit:
Fixed the PCRE download location

markos committed:
Merge pull request #113 from danlark1/develop

Optimize vectorscan for aarch64 by using shrn instruction

markos committed:
Merge pull request #116 from pareenaverma/develop

Fixed the PCRE download location

hongyang7 committed:
chimera: fix SKIP flag issue

fix github issue #360

hongyang7 committed:
UTF-8 validation: fix one cotec check corner issue

fix github issue #362

hongyang7 committed:
update year 2022

hongyang7 committed:
Fix cmake CMP0115 warning for CMake 3.20 and above

hongyang7 committed:
Silence clang-14 warnings

markos committed:
Add missing <memory> header

markos committed:
add popcount32x4, popcount64x4 helper functions

markos committed:
Merge pull request #118 from VectorCamp/bugfix/hyperscan-backport-202208

Bugfix/hyperscan backport 202208

markos committed:
[VSX] optimized mask1bit128(), moved simd_onebit_masks to common

markos committed:
[VSX] optimize and correct lshift_m128/rshift_m128

markos committed:
[VSX] huge optimization of movemask128

markos committed:
[VSX] correct lshiftbyte_m128/rshiftbyte_m128, variable_byte_shift

markos committed:
[VSX] add algorithm for alignr w/o use of immediates

markos committed:
optimize comparemask implementation, clean up code, use union types instead of casts

markos committed:
[VSX] optimize shifting methods, replace template Unroller

markos committed:
[VSX] optimize shift operators

markos committed:
[VSX] optimize alignr method

markos committed:
remove simd_onebit_masks from arm/x86 headers, as they moved to common

markos committed:
[NEON] optimize mask1bit128, get rid of simd_onebit_masks

markos committed:
readd simd_onebit_masks for x86, needs more work

markos committed:
provide non-immediate versions of lshiftbyte/rshiftbyte on x86

markos committed:
use correct intrinsic for lshiftbyte_m128

markos committed:
Merge pull request #119 from VectorCamp/feature/vsx-optimizations

VSX optimizations

liquidaty committed:
fix to enable successful build with mingw64

markos committed:
Merge pull request #121 from liquidaty/mingw64-develop

fix to enable successful build with mingw64

markos committed:
[NEON] simplify/optimize shift/align primitives

markos committed:
Merge pull request #123 from VectorCamp/feature/neon-shift-optimizations

[NEON] simplify/optimize shift/align primitives

markos committed:
Merge pull request #124 from VectorCamp/develop

Merge develop to master

markos committed:
bump version

abondarev84 committed:
fix to correctly place the autodetected flags and to activate SVE options

abondarev84 committed:
updated README to reflect CMake changes

abondarev84 committed:
removed cpu reference flags and fixed tune flag

abondarev84 committed:
additional mcpu flag cleanup

markos committed:
Jenkins change envVars -> withEnv

markos committed:
Declarative Pipeline Jenkins environment attempt

markos committed:
Declarative Pipeline Jenkins environment

markos committed:
[VSX] clang complains about the order of __vector

markos committed:
env vars have to be in quotes

markos committed:
[VSX] movemask needs to be explicitly aligned on clang for vec_ste

markos committed:
clang 14 complains about this, needs investigation

markos committed:
clang 14 does not allow bitwise OR for bools

markos committed:
remove leftover debug print

markos committed:
clang 13+ gives wrong -Wunused-but-set-variable error on nfa/mcclellancompile.cpp about total_daddy variable, disabling

markos committed:
move variable

markos committed:
Merge pull request #129 from VectorCamp/bugfix/fix-clang-on-power

Fix compile errors on clang and Power

abondarev84 committed:
SVE enabled on user input. updated README

tune and arch flags will be applied from autodetect only if they have been created by the process, otherwise the old logical flow remains wrt the flags

abondarev84 committed:
clang SVE build fix

markos committed:
Merge pull request #125 from abondarev84/master

cmake change for correct placement of autodetected tune & arch flags of GCC and SVE enablement on AARCH64

markos committed:
Merge pull request #131 from VectorCamp/develop

Prepare for new release 5.4.9

hongyang7 committed:
stream close: free stream to avoid memory leak

fix github issue #303

hongyang7 committed:
fix nfa dump error

hongyang7 committed:
scratch: add quick validity check

fix github issue #350

markos committed:
prefix assume_aligned to avoid clash with std::assume_aligned in c++20

markos committed:
Merge pull request #136 from VectorCamp/feature/prefix-assume-aligned

prefix assume_aligned to avoid clash with std::assume_aligned in c++20

fatchanghao committed:
changelog: updates for 5.4.1 release

hongyang7 committed:
scratch: remove quick validity check
Roll back fix for github issue #350

About Scratch Usage:
For compile time, scratch space is strongly recommended to be
allocated immediately after database generation.
For runtime, besides using scratch for corresponding database,
Hyperscan also allows user to use larger scratch space allocated
for another database.
When multiple concurrent threads need to use the same databases
and a new scratch space is required, cloning the largest one is
always safe. This is realized based on API hs_scratch_size() and
hs_clone_scratch().
Behaviors beyond above are discouraged and results are undefined.

markos committed:
Set Ragel.rl char type to unsigned, #135

markos committed:
fix typo

markos committed:
Force -funsigned-char to RAGEL_C_FLAGS

markos committed:
add -funsigned-char to RAGEL_C_FLAGS, move util build after that

markos committed:
fix ExpressionParser.cpp path

markos committed:
clang 14 makes some test failed because val is uninitialized

hongyang7 committed:
remove invalid nfa dump info

markos committed:
Merge pull request #141 from VectorCamp/bugfix/hs-flag-utf8-signed-char-on-arm

Set Ragel.rl char type to unsigned, #135

markos committed:
Bump version

markos committed:
Merge pull request #142 from VectorCamp/feature/bump-version

Bump version

markos committed:
Merge pull request #143 from VectorCamp/develop

Prepare for new release 5.4.9

rschu1ze committed:
Fix compilation with libcxx 16

After upgrading our (ClickHouse's) libcxx from 15 to 16, the compiler
started to complain about usage of an incomplete type "RoseInstruction"
in this (header) function:

  void RoseProgram::replace(Iter it, std::unique_ptr<RoseInstruction> ri) {
    ...

The reason is that libcxx 16 is the first version which implements C++23
constexpr std::unique_ptr (P2273R3, see (*)). RoseProgram::replace()
happens to be be const-evaluatable and the compiler tries to run
std::unique_ptr's ctor + dtor. This fails because at this point
RoseInstruction isn't defined yet.

There are two ways of fixing this:
1. Include rose_build_instruction.h (which contains RoseInstruction)
   into rose_build_program.h. Disadvantage: The new include will
   propagate transitively into all callers.
2. Move the function implementation into the source file which sees
   RoseInstruction's definition already. Disadvantage: Template
   instantiation is no longer automatic, instead there must be either a)
   explicit template instantiation (e.g. in rose_build_program.cpp) or
   b) all callers which instantiate the function must live in the same
   source file and do the instantiations by themselves. Fortunately, the
   latter is the case here, but potential future code outside
   rose_build_program.cpp will require ugly explicit instantiation.

(*) https://en.cppreference.com/w/cpp/23

markos committed:
Merge pull request #144 from rschu1ze/rs/fix-libcxx16

Fix compilation with libcxx 16

hongyang7 committed:
changelog: updates for 5.4.2 release

azat committed:
Fix use-of-uninitialized-value due to getData128()

When temporary buffer is used in getData128(), then it may return
uninitialized data.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Use std::vector instead of boost::container::small_vector under MSan

There are some issues with dtors in boost::container::small_vector
and/or vector, which is reported by MSan as an error.

The suppression __attribute__((no_sanitize_memory)) works until
clang-15, but since clang-16 it does not. It looks like before clang-16
this no_sanitize_memory works for all child functions, while since
clang-16 only for this function. I've tried to add few others, but a) it
looks icky b) I haven't managed to finish this process.

Also I've measured the performance and it hadn't been changed. Though
boost::small_vector should be faster then std::vector, but apparently my
particular case hadn't affected too much.

And one more thing, MSan reports this only with -O0, with -O3 - it is
not reproduced.

<details>

<summary>MSan report:</summary>

_Note: it was slightly trimmed_

```
==11364==WARNING: MemorySanitizer: use-of-uninitialized-value
2023.05.10 15:40:53.000233 [ 11620 ] {} <Trace> AsynchronousMetrics: MemoryTracking: was 1012.32 MiB, peak 1012.32 MiB, free memory in arenas 0.00 B, will set to 1015.82 MiB (RSS), difference: 3.50 MiB
    0 0x55558d13289f in boost::container::vector_alloc_holder<boost::container::small_vector_allocator<std::__1::pair<unsigned char, unsigned char>, std::__1::allocator<void>, void>, unsigned long, boost::move_detail::integral_constant<unsigned int, 1u>>::deallocate(std::__1::pair<unsigned char, unsigned char>* const&, unsigned long) .cmake-llvm16-msan/./contrib/boost/boost/container/vector.hpp:455:7
    1 0x55558d139e8e in boost::container::vector_alloc_holder<boost::container::small_vector_allocator<std::__1::pair<unsigned char, unsigned char>, std::__1::allocator<void>, void>, unsigned long, boost::move_detail::integral_constant<unsigned int, 1u>>::~vector_alloc_holder() .cmake-llvm16-msan/./contrib/boost/boost/container/vector.hpp:420:16
    2 0x55558d139e0b in boost::container::vector<std::__1::pair<unsigned char, unsigned char>, boost::container::small_vector_allocator<std::__1::pair<unsigned char, unsigned char>, std::__1::allocator<void>, void>, void>::~vector() .cmake-llvm16-msan/./contrib/boost/boost/container/vector.hpp:1141:4
    3 0x55558d12a4fa in boost::container::small_vector_base<std::__1::pair<unsigned char, unsigned char>, std::__1::allocator<std::__1::pair<unsigned char, unsigned char>>, void>::~small_vector_base() .cmake-llvm16-msan/./contrib/boost/boost/container/small_vector.hpp:445:80
    4 0x55558d12a4fa in boost::container::small_vector<std::__1::pair<unsigned char, unsigned char>, 1ul, std::__1::allocator<std::__1::pair<unsigned char, unsigned char>>, void>::~small_vector() .cmake-llvm16-msan/./contrib/boost/boost/container/small_vector.hpp:564:7
    5 0x55558d13a21b in std::__1::__tuple_leaf<0ul, boost::container::small_vector<std::__1::pair<unsigned char, unsigned char>, 1ul, std::__1::allocator<std::__1::pair<unsigned char, unsigned char>>, void>, false>::~__tuple_leaf() .cmake-llvm16-msan/./contrib/llvm-project/libcxx/include/tuple:265:7
    6 0x55558d13a13a in std::__1::__tuple_impl<>::~__tuple_impl() .cmake-llvm16-msan/./contrib/llvm-project/libcxx/include/tuple:451:37
    7 0x55558d13a05b in std::__1::tuple<>::~tuple() .cmake-llvm16-msan/./contrib/llvm-project/libcxx/include/tuple:538:28
    8 0x55558d139f7b in ue2::flat_detail::flat_base<>::~flat_base() .cmake-llvm16-msan/./contrib/vectorscan/src/util/flat_containers.h:89:7
    9 0x55558d1299da in ue2::flat_set<>::~flat_set() .cmake-llvm16-msan/./contrib/vectorscan/src/util/flat_containers.h:152:7
    10 0x55558d4e4dda in ue2::(anonymous namespace)::DAccelScheme::~DAccelScheme() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:301:8
    11 0x55558d4ff6cf in void boost::container::allocator_traits<>::priv_destroy<ue2::(anonymous namespace)::DAccelScheme>(boost::move_detail::integral_constant<bool, false>, boost::container::small_vector_allocator<ue2::(anonymous namespace)::DAccelScheme, boost::container::new_allocator<void>, void>&, ue2::(anonymous namespace)::DAccelScheme*) .cmake-llvm16-msan/./contrib/boost/boost/container/allocator_traits.hpp:403:11
    12 0x55558d4fefde in void boost::container::allocator_traits<>::destroy<ue2::(anonymous namespace)::DAccelScheme>(boost::container::small_vector_allocator<ue2::(anonymous namespace)::DAccelScheme, boost::container::new_allocator<void>, void>&, ue2::(anonymous namespace)::DAccelScheme*) .cmake-llvm16-msan/./contrib/boost/boost/container/allocator_traits.hpp:331:7
    13 0x55558d4fc364 in boost::container::dtl::disable_if_trivially_destructible<>::type boost::container::destroy_alloc_n<>(boost::container::small_vector_allocator<ue2::(anonymous namespace)::DAccelScheme, boost::container::new_allocator<void>, void>&, ue2::(anonymous namespace)::DAccelScheme*, unsigned long) .cmake-llvm16-msan/./contrib/boost/boost/container/detail/copy_move_algo.hpp:988:7
    14 0x55558d517962 in boost::container::vector<>::~vector() .cmake-llvm16-msan/./contrib/boost/boost/container/vector.hpp:1138:7
    15 0x55558d4f724d in boost::container::small_vector_base<>::~small_vector_base() .cmake-llvm16-msan/./contrib/boost/boost/container/small_vector.hpp:445:80
    16 0x55558d4f724d in boost::container::small_vector<>::~small_vector() .cmake-llvm16-msan/./contrib/boost/boost/container/small_vector.hpp:564:7
    17 0x55558d4f2ff3 in ue2::findDoubleBest() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:444:1
    18 0x55558d4f2f41 in ue2::findDoubleBest() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:442:9
    19 0x55558d4f2f41 in ue2::findDoubleBest() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:442:9
    20 0x55558d4f2f41 in ue2::findDoubleBest() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:442:9
    21 0x55558d4f2f41 in ue2::findDoubleBest() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:442:9
    22 0x55558d4f2f41 in ue2::findDoubleBest() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:442:9
    23 0x55558d4f2f41 in ue2::findDoubleBest() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:442:9
    24 0x55558d4f2f41 in ue2::findDoubleBest() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:442:9
    25 0x55558d4e4af5 in ue2::findBestDoubleAccelScheme() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:556:5
    26 0x55558d4e2659 in ue2::findBestAccelScheme() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:569:27
    27 0x55558d3aa8ff in ue2::look_for_offset_accel(ue2::raw_dfa const&, unsigned short, unsigned int) .cmake-llvm16-msan/./contrib/vectorscan/src/nfa/accel_dfa_build_strat.cpp:197:22
    28 0x55558d3a9727 in ue2::accel_dfa_build_strat::find_escape_strings(unsigned short) const .cmake-llvm16-msan/./contrib/vectorscan/src/nfa/accel_dfa_build_strat.cpp:414:13
    29 0x55558d3b2119 in ue2::accel_dfa_build_strat::getAccelInfo(ue2::Grey const&)::$_0::operator()(unsigned long) const .cmake-llvm16-msan/./contrib/vectorscan/src/nfa/accel_dfa_build_strat.cpp:606:26
    30 0x55558d3aefd4 in ue2::accel_dfa_build_strat::getAccelInfo(ue2::Grey const&) .cmake-llvm16-msan/./contrib/vectorscan/src/nfa/accel_dfa_build_strat.cpp:627:13
    31 0x55558d2fc61f in ue2::mcclellanCompile8() .cmake-llvm16-msan/./contrib/vectorscan/src/nfa/mcclellancompile.cpp:935:22
    32 0x55558d2e89ec in ue2::mcclellanCompile_i() .cmake-llvm16-msan/./contrib/vectorscan/src/nfa/mcclellancompile.cpp:1510:15
    33 0x55558d2ff502 in ue2::mcclellanCompile() .cmake-llvm16-msan/./contrib/vectorscan/src/nfa/mcclellancompile.cpp:1527:12
    34 0x55558fb13b52 in ue2::getDfa() .cmake-llvm16-msan/./contrib/vectorscan/src/rose/rose_build_bytecode.cpp:646:15
    35 0x55558fb7e8c8 in ue2::makeLeftNfa() .cmake-llvm16-msan/./contrib/vectorscan/src/rose/rose_build_bytecode.cpp:854:22
    36 0x55558fb6bd36 in ue2::buildLeftfix() .cmake-llvm16-msan/./contrib/vectorscan/src/rose/rose_build_bytecode.cpp:1123:15
    37 0x55558fb21020 in ue2::buildLeftfixes() .cmake-llvm16-msan/./contrib/vectorscan/src/rose/rose_build_bytecode.cpp:1579:9
    38 0x55558fad972c in ue2::buildNfas() .cmake-llvm16-msan/./contrib/vectorscan/src/rose/rose_build_bytecode.cpp:2063:10
    39 0x55558fac9843 in ue2::RoseBuildImpl::buildFinalEngine(unsigned int) .cmake-llvm16-msan/./contrib/vectorscan/src/rose/rose_build_bytecode.cpp:3660:10
    40 0x55558f2b2d86 in ue2::RoseBuildImpl::buildRose(unsigned int) .cmake-llvm16-msan/./contrib/vectorscan/src/rose/rose_build_compile.cpp:1796:12

  Uninitialized value was stored to memory at
    0 0x55558d132898 in boost::container::vector_alloc_holder<boost::container::small_vector_allocator<std::__1::pair<unsigned char, unsigned char>, std::__1::allocator<void>, void>, unsigned long, boost::move_detail::integral_constant<unsigned int, 1u>>::deallocate(std::__1::pair<unsigned char, unsigned char>* const&, unsigned long) .cmake-llvm16-msan/./contrib/boost/boost/container/vector.hpp:455:56
    1 0x55558d139e8e in boost::container::vector_alloc_holder<>::~vector_alloc_holder() .cmake-llvm16-msan/./contrib/boost/boost/container/vector.hpp:420:16
    2 0x55558d139e0b in boost::container::vector<>::~vector() .cmake-llvm16-msan/./contrib/boost/boost/container/vector.hpp:1141:4
    3 0x55558d12a4fa in boost::container::small_vector_base<>::~small_vector_base() .cmake-llvm16-msan/./contrib/boost/boost/container/small_vector.hpp:445:80
    4 0x55558d12a4fa in boost::container::small_vector<std::__1::pair<unsigned char, unsigned char>, 1ul, std::__1::allocator<std::__1::pair<unsigned char, unsigned char>>, void>::~small_vector() .cmake-llvm16-msan/./contrib/boost/boost/container/small_vector.hpp:564:7
    5 0x55558d13a21b in std::__1::__tuple_leaf<>::~__tuple_leaf() .cmake-llvm16-msan/./contrib/llvm-project/libcxx/include/tuple:265:7
    6 0x55558d13a13a in std::__1::__tuple_impl<>::~__tuple_impl .cmake-llvm16-msan/./contrib/llvm-project/libcxx/include/tuple:451:37
    7 0x55558d13a05b in std::__1::tuple<>::~tuple() .cmake-llvm16-msan/./contrib/llvm-project/libcxx/include/tuple:538:28
    8 0x55558d139f7b in ue2::flat_detail::flat_base<>::~flat_base() .cmake-llvm16-msan/./contrib/vectorscan/src/util/flat_containers.h:89:7
    9 0x55558d1299da in ue2::flat_set<>::~flat_set() .cmake-llvm16-msan/./contrib/vectorscan/src/util/flat_containers.h:1
52:7
    10 0x55558d4e4dda in ue2::(anonymous namespace)::DAccelScheme::~DAccelScheme() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:301:8
    11 0x55558d4ff6cf in void boost::container::allocator_traits<>::priv_destroy<ue2::(anonymous namespace)::DAccelScheme>() .cmake-llvm16-msan/./contrib/boost/boost/container/allocator_traits.hpp:403:11
    12 0x55558d4fefde in void boost::container::allocator_traits<>::destroy<ue2::(anonymous namespace)::DAccelScheme>(boost::container::small_vector_allocator<>&, ue2::(anonymous namespace)::DAccelScheme*) .cmake-llvm16-msan/./contrib/boost/boost/container/allocator_traits.hpp:331:7
    13 0x55558d4fc364 in boost::container::dtl::disable_if_trivially_destructible<>::type boost::container::destroy_alloc_n<boost::container::small_vector_allocator<ue2::(anonymous namespace)::DAccelScheme, boost::container::new_allocator<void>, void>, ue2::(anonymous namespace)::DAccelScheme*, unsigned long>(boost::container::small_vector_allocator<ue2::(anonymous namespace)::DAccelScheme, boost::container::new_allocator<void>, void>&, ue2::(anonymous namespace)::DAccelScheme*, unsigned long) .cmake-llvm16-msan/./contrib/boost/boost/container/detail/copy_move_algo.hpp:988:7
    14 0x55558d517962 in boost::container::vector<ue2::(anonymous namespace)::DAccelScheme, boost::container::small_vector_allocator<ue2::(anonymous namespace)::DAccelScheme, boost::container::new_allocator<void>, void>, void>::~vector() .cmake-llvm16-msan/./contrib/boost/boost/container/vector.hpp:1138:7
    15 0x55558d4f724d in boost::container::small_vector_base<>::~small_vector_base() .cmake-llvm16-msan/./contrib/boost/boost/container/small_vector.hpp:445:80
    16 0x55558d4f724d in boost::container::small_vector<>::~small_vector() .cmake-llvm16-msan/./contrib/boost/boost/container/small_vector.hpp:564:7
    17 0x55558d4f2ff3 in ue2::findDoubleBest() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:444:1
    18 0x55558d4f2f41 in ue2::findDoubleBest() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:442:9
    19 0x55558d4f2f41 in ue2::findDoubleBest() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:442:9
    20 0x55558d4f2f41 in ue2::findDoubleBest() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:442:9
    21 0x55558d4f2f41 in ue2::findDoubleBest() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:442:9

  Member fields were destroyed
    0 0x5555652e08dd in __sanitizer_dtor_callback_fields /src/llvm/worktrees/llvm-16/compiler-rt/lib/msan/msan_interceptors.cpp:961:5
    1 0x55558d4f71a6 in boost::container::small_vector<>::~small_vector() .cmake-llvm16-msan/./contrib/boost/boost/container/small_vector.hpp:528:8
    2 0x55558d4f71a6 in boost::container::small_vector<>::~small_vector() .cmake-llvm16-msan/./contrib/boost/boost/container/small_vector.hpp:564:7
    3 0x55558d4f2ff3 in ue2::findDoubleBest() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:444:1
    4 0x55558d4f2f41 in ue2::findDoubleBest() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:442:9
    5 0x55558d4f2f41 in ue2::findDoubleBest() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:442:9
    6 0x55558d4f2f41 in ue2::findDoubleBest() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:442:9
    7 0x55558d4f2f41 in ue2::findDoubleBest() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:442:9
    8 0x55558d4f2f41 in ue2::findDoubleBest() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:442:9
    9 0x55558d4f2f41 in ue2::findDoubleBest() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:442:9
    10 0x55558d4f2f41 in ue2::findDoubleBest() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:442:9
    11 0x55558d4e4af5 in ue2::findBestDoubleAccelScheme() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:556:5
    12 0x55558d4e2659 in ue2::findBestAccelScheme() .cmake-llvm16-msan/./contrib/vectorscan/src/nfagraph/ng_limex_accel.cpp:569:27
    13 0x55558d3aa8ff in ue2::look_for_offset_accel(ue2::raw_dfa const&, unsigned short, unsigned int) .cmake-llvm16-msan/./contrib/vectorscan/src/nfa/accel_dfa_build_strat.cpp:197:22
    14 0x55558d3a9727 in ue2::accel_dfa_build_strat::find_escape_strings(unsigned short) const .cmake-llvm16-msan/./contrib/vectorscan/src/nfa/accel_dfa_build_strat.cpp:414:13
    15 0x55558d3b2119 in ue2::accel_dfa_build_strat::getAccelInfo(ue2::Grey const&)::$_0::operator()(unsigned long) const .cmake-llvm16-msan/./contrib/vectorscan/src/nfa/accel_dfa_build_strat.cpp:606:26
    16 0x55558d3aefd4 in ue2::accel_dfa_build_strat::getAccelInfo(ue2::Grey const&) .cmake-llvm16-msan/./contrib/vectorscan/src/nfa/accel_dfa_build_strat.cpp:627:13
    17 0x55558d2fc61f in ue2::mcclellanCompile8(ue2::(anonymous namespace)::dfa_info&, ue2::CompileContext const&, std::__1::set<unsigned short, std::__1::less<unsigned short>, std::__1::allocator<unsigned short>>*) .cmake-llvm16-msan/./contrib/vectorscan/src/nfa/mcclellancompile.cpp:935:22
    18 0x55558d2e89ec in ue2::mcclellanCompile_i(ue2::raw_dfa&, ue2::accel_dfa_build_strat&, ue2::CompileContext const&, bool, std::__1::set<unsigned short, std::__1::less<unsigned short>, std::__1::allocator<unsigned short>>*) .cmake-llvm16-msan/./contrib/vectorscan/src/nfa/mcclellancompile.cpp:1510:15
    19 0x55558d2ff502 in ue2::mcclellanCompile(ue2::raw_dfa&, ue2::CompileContext const&, ue2::ReportManager const&, bool, bool, std::__1::set<unsigned short, std::__1::less<unsigned short>, std::__1::allocator<unsigned short>>*) .cmake-llvm16-msan/./contrib/vectorscan/src/nfa/mcclellancompile.cpp:1527:12
    20 0x55558fb13b52 in ue2::getDfa(ue2::raw_dfa&, bool, ue2::CompileContext const&, ue2::ReportManager const&) .cmake-llvm16-msan/./contrib/vectorscan/src/rose/rose_build_bytecode.cpp:646:15
```

</details>

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

markos committed:
Merge pull request #148 from azat-ch/getData128-msan

Fix use-of-uninitialized-value due to getData128()

markos committed:
Merge pull request #149 from azat-ch/small-vector-msan

Use std::vector instead of boost::container::small_vector under MSan

jeffplaisance committed:
adding ifndef around HS_PUBLIC_API definition so that vectorscan can be statically linked into another shared library without exporting symbols

Commit:
initial attempt for fat binary on Aarch64

Commit:
fix SVE flags detection order #145

markos commented:
Hi, thanks for your port to Loongson! I admit you beat me to it, I was looking forward to doing this myself! 
This looks like a complete job and I am definitely going to check it out. I already have the hardware to build locally (a Loongson 7a2000 3a5000 board) but I have yet to configure it with Debian first. I am working on some other changes in the meantime on Arm first so it might be that you have to make a few modifications to this PR still. In the meantime would you please change this PR to be against develop branch instead of master?

Thank you again!

xiangzhai commented:
Hi @markos 

> I admit you beat me to it, I was looking forward to doing this myself!

You are welcome!

> I already have the hardware to build locally (a Loongson 7a2000 3a5000 board) but I have yet to configure it with Debian first.

Yes, I am using the same working environment also for [OpenJDK](https://github.com/loongson/jdk) porting owing to I am a [JDK 8 Updates Project Author](https://openjdk.java.net/census#lzhai).

```
Architecture:        loongarch64
Byte Order:          Little Endian
CPU(s):              4
On-line CPU(s) list: 0-3
Thread(s) per core:  1
Core(s) per socket:  4
Socket(s):           1
NUMA node(s):        1
CPU family:          Loongson-64bit
Model name:          Loongson-3A5000-HV
BogoMIPS:            5000.00
L1d cache:           64K
L1i cache:           64K
L2 cache:            256K
L3 cache:            16384K
NUMA node0 CPU(s):   0-3
Flags:               cpucfg lam ual fpu lsx lasx complex crypto lvz lbt_x86 lbt_arm lbt_mips
```

And I am using LLVM toolchain for building vectorscan which depends on C++17 support.

```
$ sudo apt-get install clang-13
```

And there is `/usr/lib/llvm-13/lib/clang/13.0.1/include/lsxintrin.h`on my [Loongnix Desktop](http://pkg.loongnix.cn/loongnix/isos/Loongnix-20.5/).

```
$ mkdir build
$ cd build
$ cmake .. -DCMAKE_C_COMPILER=clang-13 -DCMAKE_CXX_COMPILER=clang++-13
```

> I am working on some other changes in the meantime on Arm first so it might be that you have to make a few modifications to this PR still.

OK. I will update my patch.

> In the meantime would you please change this PR to be against develop branch instead of master?

OK. Could you recommend which develop branch to rebase, for example, [enable-fat-runtime-arm](https://github.com/VectorCamp/vectorscan/tree/feature/enable-fat-runtime-arm)?

Thanks,
Leslie Zhai

markos commented:
> OK. Could you recommend which develop branch to rebase, for example, [enable-fat-runtime-arm](https://github.com/VectorCamp/vectorscan/tree/feature/enable-fat-runtime-arm)?

That's exactly the one, thanks again for your work!

xiangzhai commented:
> > OK. Could you recommend which develop branch to rebase, for example, [enable-fat-runtime-arm](https://github.com/VectorCamp/vectorscan/tree/feature/enable-fat-runtime-arm)?
> 
> That's exactly the one, thanks again for your work!

Rebased!

Thanks,
Leslie Zhai

xiangzhai commented:
Hi @markos 

Do I need to recreate a new PR for merging into [VectorCamp:feature/enable-fat-runtime-arm](https://github.com/VectorCamp/vectorscan/tree/feature/enable-fat-runtime-arm)?

Thanks,
Leslie Zhai

markos commented:
No, you just need to change the target of the PR, but please hold on, I'm going to merge this soon to develop, so you can just edit this one. Give me a couple of days.

markos committed:
Merge pull request #165 from VectorCamp/feature/enable-fat-runtime-arm

Feature/enable fat runtime arm

markos committed:
Merge pull request #164 from jeffplaisance/develop

adding ifndef around HS_PUBLIC_API definition so that vectorscan can be statically linked into another shared library without exporting symbols

markos committed:
Merge pull request #169 from VectorCamp/feature/backport-hyperscan-2023Q3

Feature/backport hyperscan 2023 q3

markos committed:
bump version, add Vectorscan Changelog

markos committed:
minor fix

markos committed:
Merge pull request #167 from VectorCamp/develop

Prepare for 5.4.10

markos committed:
forgot to update changelog for latest entry

xiangzhai commented:
Hi @markos 

I rebased the initial loongarch port patch.

Thanks,
Leslie Zhai

markos committed:
force disable FAT_RUNTIME on MacOS on Arm

markos committed:
HWCAP is only available on Linux

markos committed:
Merge pull request #173 from VectorCamp/bugfix/disable-fat-macos-arm

Bugfix/disable fat macos arm

markos committed:
Merge pull request #174 from VectorCamp/develop

Minor bugfix release 5.4.10.1

markos commented:
Hi @xiangzhai, I have finally setup my 3a5000 system and I was able to attempt to build your Pull Request. Unfortunately, the compiler tested (gcc 13 on Debian and Archlinux) do not hold this header: lsxintrin.h. So the build failed really early. After some research I realized that it's quite a bit early in terms of compiler support. Until this is finalized, I'm afraid I cannot test nor approve this Pull Request and it will have to wait. We can certainly revisit later of course, now that I have a working system I'm going to test this frequently for updates.
Again apologies for this, but even though I can tell it's high quality code you wrote, the actual environment is still in the early experimental phase and it's untestable.

xiangzhai commented:
Hi @markos 

> the compiler tested (gcc 13 on Debian and Archlinux) do not hold this header: lsxintrin.h

Yes, `/usr/lib/llvm-13/lib/clang/13.0.1/include/lsxintrin.h` is only available for [Loongnix Desktop](http://pkg.loongnix.cn/loongnix/isos/Loongnix-20.5/) and [Loongnix Server](http://pkg.loongnix.cn/loongnix-server/8.4/isos/).

Let's wait [upstream LSX/LASX support](https://lists.llvm.org/pipermail/llvm-dev/2021-December/154371.html) for LLVM toolchain. The reviewers are also code review the LLVM and GCC patchset for toolchains :)

> Again apologies for this, but even though I can tell it's high quality code you wrote, the actual environment is still in the early experimental phase and it's untestable.

You are welcome!

Thanks,
Leslie Zhai

markos committed:
Fix version getting out of sync #175

markos committed:
Fix 'unqualified call to std::move' errors in clang 15+

markos committed:
move definition of RAGEL_C_FLAGS earlier to catch tools/hscollider

markos committed:
more std::move fixes

Commit:
fix -Wunused warnings on debug

markos committed:
Reduce unit test runtimes dramatically for debug builds

markos committed:
clang 15 (but not 16) fails on ppc64le with -Wdeprecate-lax-vec-conv-all

markos committed:
add conditional for __clang__

markos committed:
use the conditional in the right way

markos committed:
missed one pragma

markos committed:
clang 16 as well

markos committed:
use the right type of cast

markos committed:
Don't run regression UE_2595 on debug, it times out CI

markos committed:
Reduce debug unit tests runtime even more
In single.cpp featuremask with AVX512 features is not relevant to non-x86 platforms,
and just extends the runtime for no reason.

markos committed:
__builtin_constant_p is true in the wrong case on gcc 13.2. Exclude for now

markos committed:
remove extra print

markos committed:
detect arm_sve.h when using clang on fat runtime builds

markos committed:
fix fat & normal build errors on arm

markos committed:
re-add missing file to x86 builds

markos committed:
add cpuid_flags to ppc64le as well

markos committed:
add missing file

markos committed:
include extra sources for Arm on non-fat builds

markos committed:
add src/nfa/vermicelli_simd.cpp to ppc64le

markos committed:
SVE2 needs armv9-a, fix build

markos committed:
WIP: Refactor CMake build system to more modular

markos committed:
add missing file

markos committed:
increase cmake_minimum_version, basically the one in Debian 11

markos committed:
fix defaults for -march for x86

markos committed:
fix cflags detection for x86

markos committed:
fix sqlite3 version detection

markos committed:
fix x86 fat binary build

markos committed:
fix default arch/tune flags for ppc64le

Commit:
fix cmake refactor for arm builds

markos committed:
is not known at that stage

markos committed:
remove unneeded option

markos committed:
remove vermicelli_simd.cpp to fix redefinition build failure on SVE2 builds

markos committed:
SVE2 is armv9-a but gcc 11 does not recognize that

markos committed:
set default value

markos committed:
fix default arch definition for non fat builds on arm

markos committed:
Merge pull request #181 from VectorCamp/bugfix/fix-clang15-compilation-errors

Fix clang 15,16 compilation errors on all platforms, refactor CMake build system

markos committed:
Ubuntu 20.04 gcc does not define HWCAP2_SVE2 #180

markos commented:
could you please set this PR against develop branch? Various changes were made in the CMake build system and this PR will have to be refactored.

xiangzhai commented:
Hi @markos 

> could you please set this PR against develop branch? Various changes were made in the CMake build system and this PR will have to be refactored.

Rebased! Passed unit-internal and unit-hyperscan tests.

Thanks,
Leslie Zhai

markos committed:
Merge pull request #185 from VectorCamp/bugfix/fix-inconsistent-version-header

Fix version getting out of sync #175

markos committed:
Merge pull request #186 from VectorCamp/bugfix/fix-compilation-arm-ubuntu-20.04

Ubuntu 20.04 gcc does not define HWCAP2_SVE2 #180

Commit:
make pkgconfig a requirement

mlmitch committed:
Correct set_source_files_properties usage

The use of `CMAKE_BINARY_DIR` and `CMAKE_CURRENT_BINARY_DIR` when
specifying files to set_source_files_properties caused problems
when this project is used from another CMake project.

More specifically, these variables aren't set to the expected path,
and the properties are attempted to be set for non-existant files.

This was benign before vectorscan 5.4.8 as the only properties
set were warning suppression flags.

Starting with 5.4.9, `-funsigned-char` was applied to Ragel outputs
using this method. The result is projects depending on Vectorscan
through Cmake do not have this compile flag properly applied.

markos committed:
Merge pull request #188 from VectorCamp/bugfix/require-pkg-config

make pkgconfig a requirement

markos committed:
Merge pull request #189 from mlmitch/develop

Correct set_source_files_properties usage

xiangzhai commented:
Hi @markos

> > the compiler tested (gcc 13 on Debian and Archlinux) do not hold this header: lsxintrin.h
> 
> Yes, `/usr/lib/llvm-13/lib/clang/13.0.1/include/lsxintrin.h` is only available for [Loongnix Desktop](http://pkg.loongnix.cn/loongnix/isos/Loongnix-20.5/) and [Loongnix Server](http://pkg.loongnix.cn/loongnix-server/8.4/isos/).
> 
> Let's wait [upstream LSX/LASX support](https://lists.llvm.org/pipermail/llvm-dev/2021-December/154371.html) for LLVM toolchain. The reviewers are also code review the LLVM and GCC patchset for toolchains :)

Upstreamed:  https://github.com/llvm/llvm-project/blob/main/clang/lib/Headers/lsxintrin.h

So you might test (clang or gcc on Debian and Archlinux) for it.

Thanks,
Leslie Zhai

markos committed:
Move VERM16 enums to the end of the list

This was causing a hard-to-track segfault with Fat Runtime on SVE2 hw,
because of the macro-based hard-coded way to calculate offsets for each
implementation. This needs a rewrite.

markos committed:
Merge pull request #191 from VectorCamp/bugfix/fix-segfault-arm-sve2

Move VERM16 enums to the end of the list

markos committed:
remove Jenkinsfile

markos committed:
Don't build fat runtime with native CPU detection

markos committed:
bump version

markos committed:
Update Readme file

markos committed:
add changelog entry

markos committed:
add contributors file

markos committed:
additions to readme

markos committed:
if none are set build static

markos committed:
cross-compiling is not tested, removed

markos committed:
more changes to readme

markos committed:
add sanitizer flags

markos committed:
expand on build-deps installation

markos committed:
Merge pull request #196 from VectorCamp/feature/prepare-5.4.11

Feature/prepare 5.4.11

gliwka committed:
fix missing hs_version.h header (closes #198)

markos committed:
fix missing installation of static libs

markos committed:
Merge pull request #199 from gliwka/fix-missing-hs-version-header

Fix missing hs_version.h header (closes #198)

markos committed:
Merge pull request #200 from VectorCamp/bugfix/install-static-libs

fix missing installation of static libs

markos committed:
Add SIMDe backend to CMake

markos committed:
add SIMDe git submodule

markos committed:
add SIMDe cmake file

markos committed:
mofidied .gitmodules

markos committed:
add SIMDe ports of simd_utils and supervector

markos committed:
enable SIMDe backend

markos committed:
add generic tune flags

markos committed:
add include dirs

markos committed:
SIMDE is a valid platform

markos committed:
fix test for SIMDe

markos committed:
add missing intrinsics for SIMDe backend

markos committed:
existing scalar implementations were incorrect -but never tested, ported from arm/ppc64le

Commit:
fix SIMDe emulation builds on Arm, add native translation from x86 for comparison

markos committed:
fix TUNE_FLAG for SIMDE_BACKEND

markos committed:
use pkg-config for SIMDe

markos committed:
fix compilation for SIMDe

Commit:
make diffrich384 available on all arches

markos committed:
add missing else

markos committed:
separate ARCH_FLAG logic

markos committed:
search for SIMDE sse4.2.h header

markos committed:
updates to the Readme

markos committed:
fix failing allbits test for ppc64le on clang15

markos committed:
Merge pull request #203 from VectorCamp/feature/enable-simde-backend

Feature/enable simde backend

xiangzhai committed:
Initial loongarch port

Co-authored-by: yangwenqing <yangwenqing@loongson.cn>

Signed-off-by: Leslie Zhai <zhaixiang@loongson.cn>
Signed-off-by: yangwenqing <yangwenqing@loongson.cn>

xiangzhai commented:
Hi @markos 

Could you review the patch?

Thanks,
Leslie Zhai

markos commented:
@xiangzhai not until lsxintrin.h is in a package in Debian for either clang or gcc. Until that happens, I cannot merge LSX support. I just tested gcc-13 and clang-17, they still don't have that file in Debian packages.

markos commented on CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required (VERSION 3.18.4)
+cmake_minimum_required (VERSION 3.13.4)
```
please don't change that.

xiangzhai replied:
Done.

markos commented on CMakeLists.txt:
```diff
@@ -160,6 +162,11 @@ foreach (CONFIG ${CMAKE_BUILD_TYPE} ${CMAKE_CONFIGURATION_TYPES})
     string(REGEX REPLACE "-O[^ ]*" "" CMAKE_CXX_FLAGS_${CONFIG} "${CMAKE_CXX_FLAGS_${CONFIG}}")
 endforeach ()
 
+if (ARCH_LOONGARCH64)
```
these do not belong here, but in their respective cmake file. Please, see the other architectures.

xiangzhai replied:
Done.

markos commented on CMakeLists.txt:
```diff
@@ -186,7 +193,7 @@ set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${ARCH_CXX_FLAGS}")
 # and hscollider
 set(PCRE_REQUIRED_MAJOR_VERSION 8)
-set(PCRE_REQUIRED_MINOR_VERSION 41)
+set(PCRE_REQUIRED_MINOR_VERSION 39)
```
Please do not change that.

xiangzhai replied:
Done.

markos commented on src/util/arch/loongarch64/match.hpp:
```diff
@@ -0,0 +1,120 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+static really_inline m128 vpmax_loongarch(v4u32 a, v4u32 b) {
```
This might be useful in other places as well, perhaps it should be added to loongarch64/simd_utils.h?

markos commented:
I added some review comments, which should be easy to fix until we can actually do a proper review when compiler support arrives. Thanks for your patience in this.

xiangzhai committed:
Revert cmake_minimum_required and PCRE_REQUIRED_MINOR_VERSION change.
Move ARCH_C_FLAGS and ARCH_CXX_FLAGS to LoongArch respective cmake file.
Move vpmax_loongarch to loongarch64/simd_utils.h.

xiangzhai commented:
Hi @markos 

Thanks for your careful review!

I updated my patch as your suggestion.

Thanks,
Leslie Zhai

xiangzhai commented:
Hi @markos 

> not until lsxintrin.h is in a package in Debian for either clang or gcc. Until that happens, I cannot merge LSX support. I just tested gcc-13 and clang-17, they still don't have that file in Debian packages.

Please update to install clang-18, there is `lsxintrin.h` for LLVM 18 now and [rust SIMD intrinsics for LoongArch merged](https://github.com/rust-lang/stdarch/pull/1535) :)

Thanks,
Leslie Zhai

markos commented:
Hi @xiangzhai 
I will have to unbrick my 3a5000 system as I tried to flash the wrong uefi image and it doesn't boot now. I will do that in the next weeks and reconfigure the system to use a more modern compiler. If I can make it work, I will merge this PR. 
Thank you again for your contribution!

xiangzhai commented:
> I will have to unbrick my 3a5000 system as I tried to flash the wrong uefi image and it doesn't boot now.

https://github.com/loongson/Firmware/tree/main/5000Series/PC

markos commented:
> > I will have to unbrick my 3a5000 system as I tried to flash the wrong uefi image and it doesn't boot now.
> 
> https://github.com/loongson/Firmware/tree/main/5000Series/PC

Indeed, that's where I got it from, I just forgot to download the 'raw' binary. In any case, I need to figure out how to connect a JTAG or similar adaptor and upload the UEFI image using SPI. Currently it just gives a constant beep and fails to do anything else -no UEFI Shell/BIOS screen or anything else.

xiangzhai commented:
Hi @markos 

Could you try [qemu system loongarch64](https://github.com/foxsen/qemu-loongarch-runenv) if your working environment was broken please?

Thanks,
Leslie Zhai

markos commented:
Even if it builds on qemu -which is not something I have the time to do- we would not merge the PR as we would not be able to support it in our CI. It is in the immediate plans to either fix the system or get a new one. Just be patient please.

markos commented:
Closing this, current hardware in our lab (3a5000) is unstable difficult to work. We cannot accept any PR for a port we will have to maintain ourselves. If hardware becomes better available we can revisit. If you need an optimized port for Loongson we will have to do the work ourselves, please contact us for a support contract.
